server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class DesiredNodesMembershipService implements ClusterStateListener {
+    private final Set<DesiredNode> members;
+    private String latestHistoryId = null;
+
+    DesiredNodesMembershipService() {
+        this.members = new HashSet<>();
+    }
+
+    public static DesiredNodesMembershipService create(ClusterService clusterService) {
+        var tracker = new DesiredNodesMembershipService();
+        clusterService.addListener(tracker);
+        return tracker;
+    }
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
+        final var clusterState = event.state();
+        final var desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        if (desiredNodes == null) {
+            final var previousDesiredNodes = DesiredNodes.latestFromClusterState(event.previousState());
+            if (previousDesiredNodes != null) {
```

Comment:
Do we need this condition? Seems more expensive than just clearing `members` always?